### ISSUE TITLE
Splitchunks and shared code between worker files using umd creates forever hanging code

### ISSUE THREAD

**Issue by onzag** [webpack/webpack#10278]
# Bug report

**What is the current behavior?**

If you have two entries, one for your worker and another one for your main file, where code is shared between those two, and if you use split chunks so that shared code is split in a different file, the worker hangs undefinetely because simply the common file cannot be loaded, and there is no way for it to be loaded.

Using import() is not a good alternative as you lose the type definitions; which means you are forced to create some null-loader for the type definitions or some custom definition file (which is really not good at all).

Also you need a hack to even make it work https://stackoverflow.com/a/59743427/1199740 which is not optimal.

Not to add it depends on Promises, and code in common, usually includes the promises polyfill, since you are within a worker that shouldn't be an issue, but still, not optimal.

**If the current behavior is a bug, please provide the steps to reproduce.**

Create a project with two entries.

Load a second entry as a worker.

It will hang forever.

**What is the expected behavior?**

1. You should be able to use of importScripts if you are inside a worker; you should specify if an entry should use importScripts to fullfill its imports.

OR

2. You should be able to inject the code required for initialization, self.importScripts("/dist/build.common.js") just the same that you would do by adding it in the DOM; eg by using an injection file instead that would be bundled but won't hang, that file would execute importScripts just as in the normal DOM available version you would create a <script> with the commons file.

AND

You shouldn't depend on the DOM for dynamic imports, if importScripts is available it should be used instead.

**Other relevant information:**
webpack version:  4.41.4
Node.js version: 12.14.0
Operating System: Ubuntu
Additional tools:

**Comment by sokra** [comment]
For WebWorker builds you need to use `target: "webworker"`. Sadly having different targets per entrypoint is not supported yet. So you need to use multiple compilations for worker and app builds, which do not support shared chunks.

Target per entrypoint is planned for webpack 5.

**Comment by onzag** [comment]
@sokra sorry for the late reply I didn't get any notification.

I don't understand, I actually managed to get this working by writing a plugin, shared code between workers and app builds is totally possible, it's just the building step that is unable.

I managed to do 2. of my possible solution npmjs.com/package/worker-injector-generator-plugin here it is what I came up with.

I don't think this is optimal but does the job, given a cache system in place the main app and the worker don't trigger the network twice.
